File: yixi/src/utils/songCoverUrl.ts
export type SongCoverSize = 120 | 240 | 360 | 480;

type SongCoverInput = {
  source?: string;
  cover?: string;
  coverSize?: unknown;
  picUrl?: string;
  coverUrl?: string;
  coverImgUrl?: string;
  imgurl?: string;
  img1v1Url?: string;
  album?: unknown;
  al?: unknown;
};

export function getKgImageUrl(url?: string, size: SongCoverSize = 120, fallback = ""): string {
  if (!url) return fallback;
  return url.replace("{size}", size.toString());
}

export function getWyCoverSizeUrl(url?: string, size: number | null = null, fallback = "") {
  try {
    if (!url) return fallback;
    const sizeUrl = size
      ? typeof size === "number"
        ? `?param=${size}y${size}`
        : `?param=${size}`
      : "";
    const imageUrl = url.replace(/^http:/, "https:");
    if (imageUrl.endsWith(".jpg")) {
      return imageUrl + sizeUrl;
    }
    if (imageUrl.endsWith("&")) {
      const nextUrl = imageUrl + "cl";
      return nextUrl.replace(
        /(thumbnail=[0-9]+y[0-9]+&cl)/,
        `thumbnail=${size}y${size}&`
      );
    }
    return imageUrl;
  } catch (error) {
    console.error("图片链接处理出错：", error);
    return fallback;
  }
}

export function getWyCoverUrl(item: SongCoverInput, fallback = "") {
  const album = asRecord(item.album);
  const al = asRecord(item.al);
  const alXInfo = asRecord(al?.xInfo);
  const cover =
    item.cover ||
    item.picUrl ||
    item.coverUrl ||
    item.coverImgUrl ||
    item.imgurl ||
    item.img1v1Url ||
    toStringValue(album?.picUrl) ||
    toStringValue(al?.picUrl) ||
    toStringValue(alXInfo?.picUrl);
  const coverSize = {
    s: getWyCoverSizeUrl(cover, 100, fallback),
    m: getWyCoverSizeUrl(cover, 300, fallback),
    l: getWyCoverSizeUrl(cover, 1024, fallback),
    xl: getWyCoverSizeUrl(cover, 1920, fallback),
  };
  return { cover: cover || fallback, coverSize };
}

export function getSongCoverUrl(song: SongCoverInput, size: SongCoverSize = 120, fallback = "") {
  if (song.source === "local") return fallback;
  if (song.source === "kg") return getKgImageUrl(song.cover, size, fallback);
  if (song.source === "wy") {
    const res = getWyCoverUrl(song, fallback);
    switch (size) {
      case 120:
        return res.coverSize.s || res.cover || fallback;
      case 240:
        return res.coverSize.m || res.cover || fallback;
      case 360:
        return res.coverSize.l || res.cover || fallback;
      case 480:
        return res.coverSize.xl || res.cover || fallback;
    }
  }
  return song.cover || fallback;
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return undefined;
}

function toStringValue(value: unknown) {
  if (typeof value === "string") return value.trim();
  if (typeof value === "number" && Number.isFinite(value)) return String(value);
  return "";
}

File: pm/app/src/main/java/cn/partialy/pm/utils/SongCoverUrl.kt
package cn.partialy.pm.utils

import cn.partialy.pm.model.SongInfo
import cn.partialy.pm.model.SongType

object SongCoverUrl {
    const val SIZE_SMALL = 120
    const val SIZE_MEDIUM = 240
    const val SIZE_LARGE = 360
    const val SIZE_XLARGE = 480

    fun getKgImageUrl(url: String?, size: Int = SIZE_SMALL, fallback: String = ""): String {
        val raw = url?.trim().orEmpty()
        if (raw.isBlank()) return fallback
        return raw.replace("{size}", normalizeSize(size).toString())
    }

    fun getSongCover(song: SongInfo, size: Int = SIZE_SMALL, fallback: String = ""): String =
        getSongCover(song.type, song.coverUrl, size, fallback)

    fun getSongCover(type: SongType, coverUrl: String?, size: Int = SIZE_SMALL, fallback: String = ""): String =
        when (type) {
            SongType.LOCAL -> fallback
            SongType.KG -> getKgImageUrl(coverUrl, size, fallback)
            SongType.WY -> getWyCoverSizeUrl(coverUrl, wySizeOf(size), fallback)
            SongType.KW -> coverUrl?.trim().orEmpty().ifBlank { fallback }
        }

    fun getSongCoverData(song: SongInfo, size: Int = SIZE_SMALL, fallback: String = ""): Any? =
        song.embeddedCoverArt ?: getSongCover(song, size, fallback).takeIf { it.isNotBlank() }

    private fun getWyCoverSizeUrl(url: String?, size: Int?, fallback: String = ""): String {
        return try {
            val raw = url?.trim().orEmpty()
            if (raw.isBlank()) return fallback
            val imageUrl = raw.replace(Regex("^http:"), "https:")
            val sizeUrl = size?.let { "?param=${it}y$it" }.orEmpty()
            when {
                imageUrl.endsWith(".jpg") -> imageUrl + sizeUrl
                imageUrl.endsWith("&") -> {
                    val nextUrl = imageUrl + "cl"
                    nextUrl.replace(
                        Regex("thumbnail=[0-9]+y[0-9]+&cl"),
                        "thumbnail=${size}y${size}&",
                    )
                }
                else -> imageUrl
            }
        } catch (_: Exception) {
            fallback
        }
    }

    private fun normalizeSize(size: Int): Int =
        when (size) {
            SIZE_MEDIUM -> SIZE_MEDIUM
            SIZE_LARGE -> SIZE_LARGE
            SIZE_XLARGE -> SIZE_XLARGE
            else -> SIZE_SMALL
        }

    private fun wySizeOf(size: Int): Int =
        when (normalizeSize(size)) {
            SIZE_SMALL -> 100
            SIZE_MEDIUM -> 300
            SIZE_LARGE -> 1024
            SIZE_XLARGE -> 1920
            else -> 100
        }
}
